src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs

```diff
@@ -457,146 +478,174 @@ public Task<ImmutableArray<Diagnostic>> GetAllDiagnosticsAsync()
+                    // Track if this task was suspended by another tree diagnostics request for the same tree.
+                    // If so, we wait for the high priority requests to complete before restarting analysis.
+                    bool suspendend;
+                    do
+                    {
+                        suspendend = false;
 
-                            cancellationToken.ThrowIfCancellationRequested();
+                        // Create a new cancellation source to allow higher priority requests to suspend our analysis.
+                        cts = new CancellationTokenSource();
 
-                            await computeTask.ConfigureAwait(false);
-                        }
-                        catch (OperationCanceledException)
-                        {
-                            cancellationToken.ThrowIfCancellationRequested();
-                            suspendend = cts.IsCancellationRequested;
-                        }
-                        finally
+                        // Link the cancellation source with client supplied cancellation source, so the public API callee can also cancel analysis.
+                        using (var linkedCts = CancellationTokenSource.CreateLinkedTokenSource(cts.Token, cancellationToken))
                         {
-                            cts.Dispose();
-                            ClearExecutingTask(computeTask, analysisScope.FilterTreeOpt);
-                            computeTask = null;
+                            try
+                            {
+                                // Core task to compute analyzer diagnostics.
+                                Func<Tuple<Task, CancellationTokenSource>> getComputeTask = () => Tuple.Create(
+                                    Task.Run(async () =>
+                                        {
+                                            try
+                                            {
+                                                AsyncQueue<CompilationEvent> eventQueue = null;
+                                                try
+                                                {
+                                                    // Get event queue with pending events to analyze.
+                                                    eventQueue = getEventQueue();
+
+                                                    // Execute analyzer driver on the given analysis scope with the given event queue.
+                                                    await ComputeAnalyzerDiagnosticsCoreAsync(driver, eventQueue, analysisScope, cancellationToken: linkedCts.Token).ConfigureAwait(false);
+                                                }
+                                                finally
+                                                {
+                                                    FreeEventQueue(eventQueue);
```

Comment:
The pattern we use for the filters is to wrap the entire method body. The goal of the filter is to intercept an exception coming from any statement of the method body before it reaches the synthesized catch-all clause that the compiler generates around the entire method body.

The final IL looks as follows:
```
try
{
  try { original body } filter { crash }  
}
catch (Exception e)      // compiler generated
{                          
   task.Exception = e;
}
```